docs/sql-ref-datatypes.md

```diff
@@ -314,3 +314,51 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+|ShortType |ShortType, IntegerType, LongType, FloatType, DoubleType|
+|IntegerType |IntegerType, LongType, FloatType, DoubleType|
+|LongType |LongType, FloatType, DoubleType|
+|FloatType |FloatType, DoubleType|
+|DoubleType |DoubleType|
+|StringType |DoubleType (in numeric operations), StringType |
+|BinaryType |BinaryType|
+|BooleanType |BooleanType|
+|TimestampType |TimestampType, DateType|
+|DateType |DateType|
+
+The case of DecimalType, is treated differently, for example, there is no common type for double and decimal because double's range is larger than decimal, and yet decimal is more precise than double, but in an operation, we would cast the decimal into double.
+
+#### Explicit casting and store assignment casting
+
+When you are using explicit casting by CAST or doing INSERT INTO operations that need to cast types to different store types, the following matrix shows if the conversion is allowed
+
+|         |ByteType  |ShortType |IntegerType |LongType |FloatType |DoubleType |StringType |BinaryType |BooleanType |TimestampType |DateType|
+|---------|----------|----------|------------|---------|----------|-----------|-----------|-----------|------------|--------------|--------|
+|ByteType |--        |X         |X           |X        |X         |X          |X          |X          |X           |X             |        |
+|ShortType|*         |--        |X           |X        |X         |X          |X          |X          |X           |X             |        |
+|IntegerType|*       |*         |--          |X        |X         |X          |X          |X          |X           |X             |        |
+|LongType |*         |*         |*           |--       |X         |X          |X          |X          |X           |X             |        |
+|FloatType |*        |*         |*           |*        |--        |X          |X          |           |X           |X             |        |
+|DoubleType |*       |*         |*           |*        |*         |--         |X          |           |X           |X             |        |
+|StringType |*       |*         |*           |*        |*         |*          |--         |X          |X           |X             |X       |
+|BinaryType |        |          |            |         |          |           |           |--         |            |              |        |
+|BooleanType |X      |X         |X           |X        |X         |X          |X          |           |--          |              |        |
+|TimestampType |*    |*         |*           |X        |X         |X          |X          |           |            |--            |X       |
+|DateType |*         |*         |X           |X        |X         |X          |X          |           |            |X             |--      |
+
+X: Conversion allowed (cast ByteType in ShortType)  
+*: An overflow can occur, check ANSI compliance for the result in this case (cast ShortType in ByteType)
```

Comment:
The overflow behaviour is different between ANSI/non-ANSI modes. Could you explain about it here and add a link to the ANSI page? https://github.com/apache/spark/blob/master/docs/sql-ref-ansi-compliance.md#type-conversion We might need a subsection for the explanation.